src/java.desktop/share/classes/javax/swing/plaf/metal/MetalFileChooserUI.java

```diff
@@ -1266,7 +1266,9 @@ protected DirectoryComboBoxAction() {
         public void actionPerformed(ActionEvent e) {
             directoryComboBox.hidePopup();
             File f = (File)directoryComboBox.getSelectedItem();
-            if (!getFileChooser().getCurrentDirectory().equals(f)) {
+            File curDir = getFileChooser().getCurrentDirectory();
+
+            if (curDir != null && !curDir.equals(f)) {
```

Comment:
If `f` is always not null, you can you use `!f.equals(curDir)`.

Reply:
> 

Prevention of not setting to null means I have to modify `JFileChooser` class itself, so just handling here. And yeah` f `can't be null, can do that.

Reply:
Why is it bad to modify `JFileChooser` class itself?

I think setting the current directory to `null` while it displays a directory kind of breaks the invariant.

`JFileChooser` navigated to a directory and the current directory is set to it. You tried navigating to another directory, `FileSystem` didn't allow the navigation, so `JFileChooser` continues to display the files in the directory that is showed before. Thus, the current directory hasn't changed. From this point of view, the current directory should remain unchanged.

Reply:
https://github.com/openjdk/jdk/blob/ccc1d3169691d066c08e294f5d989b007bfab114/src/java.desktop/share/classes/javax/swing/JFileChooser.java#L608
Here the above logic checks if the current selected directory is traversable and if not it goes up to parent directory till top level and when every check fails then null is set to `currentDirectory`. Should we think of modifying this logic to not to set to null, rather remain in previously valid folder......?

Reply:
Well, the logic there tries to prevent the current directory from becoming `null`. Yet it traverses up the tree:
https://github.com/openjdk/jdk/blob/ccc1d3169691d066c08e294f5d989b007bfab114/src/java.desktop/share/classes/javax/swing/JFileChooser.java#L603-L607

Probably, if `dir` becomes `null` in the end, it should resort to its default behaviour: `getFileSystemView().getDefaultDirectory()`. Or not change at all?

Then, you can handle the situation here: if `getFileChooser().getCurrentDirectory()` returns `null`, restore the previous current directory.

And what about other L&F? If the same code runs in Windows L&F, Nimbus L&F, any other?

Reply:
In other L&F the same problem exist, but NPE is does not occurs because previous state check is not done, directly the clicked file is been set. 
https://github.com/openjdk/jdk/blob/ccc1d3169691d066c08e294f5d989b007bfab114/src/java.desktop/windows/classes/com/sun/java/swing/plaf/windows/WindowsFileChooserUI.java#L1328

Reply:
> Well, the logic there tries to prevent the current directory from becoming `null`. Yet it traverses up the tree:
> 
> https://github.com/openjdk/jdk/blob/ccc1d3169691d066c08e294f5d989b007bfab114/src/java.desktop/share/classes/javax/swing/JFileChooser.java#L603-L607
> 
> Probably, if `dir` becomes `null` in the end, it should resort to its default behaviour: `getFileSystemView().getDefaultDirectory()`. Or not change at all?

This behaviour is specified. https://github.com/openjdk/jdk/blob/ccc1d3169691d066c08e294f5d989b007bfab114/src/java.desktop/share/classes/javax/swing/JFileChooser.java#L576-L580

Yet it doesn't take into account a custom implementation of `FileSystem` which does not allow navigating up. It may be a problem in `FileSystem` implementation.

Reply:
In other L&F, null gets assigned to currentDirectory and NPE doesn't occur since that previous directory check is not done.

Reply:
> … NPE doesn't occur since that previous directory check is not done.

Then I suggest going this way and just drop the checks from Metal L&F.

The test should set Metal L&F explicitly. Or better yet, it should be run in all L&Fs to ensure NPE isn't thrown.

<hr>

As for preserving the current value of `currentDirectory` if the passed in folder cannot be navigated to. I think it should be done yet it's a separate issue from this one. Shall I submit a new bug? Does anyone have an opinion on this?